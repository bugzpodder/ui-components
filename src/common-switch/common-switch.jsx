// @flow
import React from "react";
import Switch from "@material-ui/core/Switch";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import FormHelperText from "@material-ui/core/FormHelperText";
import classNames from "classnames";
import styles from "./common-switch.module.scss";

type Props = {
	/** Returns the status of the switch, and the value if one exists */
	onChange: (boolean, any) => any,
	/** Determines the color of the switch. Options include "primary" and "secondary" */
	color?: "primary" | "secondary",
	/** The id given to the switch */
	id?: string,
	/** When `true`, displays the switch in an error state */
	showError?: boolean,
	/** Provides a helper text under the switch */
	helperText?: string,
	/** Provides a label to the right of the switch */
	label?: string,
	/** When `false`, disables the switch */
	isEnabled?: boolean,
	/** Allows the programmer to take control of the selected state of the switch */
	isSelected?: boolean,
	/** The value of the switch input if one is provided */
	value?: string,
};

type State = {
	isChecked: boolean,
};

/** `CommonSwitch` renders a Material-UI switch component */
export class CommonSwitch extends React.Component<Props, State> {
	state = {
		isChecked: false,
	};

	handleChange = (e: InputEvent, checked: boolean) => {
		const { onChange, isSelected } = this.props;
		onChange(checked, e.currentTarget.value);
		if (isSelected === undefined || isSelected === null) {
			this.setState({ isChecked: checked });
		}
	};

	render = () => {
		const {
			id = "",
			color = "primary",
			showError = false,
			helperText = "",
			label,
			isEnabled = true,
			isSelected,
			value = "",
		} = this.props;
		const { isChecked } = this.state;
		const errorClass = {};
		errorClass[styles.error] = showError;
		const switchSelection = isSelected !== undefined || isSelected !== null ? isSelected : isChecked;
		return (
			<div className={styles.commonSwitch}>
				<FormControlLabel
					classes={{
						root: styles.controlLabel,
						label: classNames(errorClass),
					}}
					control={(
						<Switch
							id={id}
							data-testid="common-switch"
							checked={switchSelection}
							classes={{
								iconChecked: classNames(errorClass),
								colorPrimary: classNames(errorClass),
								colorSecondary: classNames(errorClass),
							}}
							color={color}
							disabled={!isEnabled}
							disableRipple
							onChange={this.handleChange}
							value={value}
						/>
)}
					label={label}
				/>
				{helperText && (
					<FormHelperText
						classes={{
							root: styles.switchHelperText,
						}}
						error={showError}
					>
						{helperText}
					</FormHelperText>
				)}
			</div>
		);
	};
}
