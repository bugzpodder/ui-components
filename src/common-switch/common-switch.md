### Example

```js
const ExampleBlock = require("../test-utils").ExampleBlock;
const ExampleWrapper = require("../test-utils").ExampleWrapper;
const styles = require("../test-utils/example-styles.module.scss");

class TestCommonSwitch extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      checkedA: false,
      checkedB: false,
      showError: false,
      value: "",
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleError = this.handleError.bind(this);
  }

  handleError(checked) {
    this.setState({ showError: checked });
  }

  handleChange(checked, val) {
    const value = checked ? val : "";
    if (val === "first") {
      this.setState({ checkedA: checked, checkedB: false, value });
      return;
    }
    this.setState({ checkedA: false, checkedB: checked, value });
  }

  render() {
    const { checkedA, checkedB, showError } = this.state;
    return (
      <div className={styles.container}>
        <CommonSwitch label="Primary" onChange={() => {}} />

        <CommonSwitch label="Secondary" color="secondary" value="secondary" onChange={() => {}} />

        <CommonSwitch
          label="Error"
          value="error"
          showError={showError}
          helperText="some helper text"
          onChange={this.handleError}
        />

        <CommonSwitch className={styles.spacing} label="Disabled" isEnabled={false} onChange={this.handleError} />
        <br />

        <CommonSwitch color="primary" value="first" isSelected={checkedA} onChange={this.handleChange} />

        <CommonSwitch value="second" isSelected={checkedB} onChange={this.handleChange} />

        <ExampleBlock strongHeader="Switch value" content={this.state.value} />
      </div>
    );
  }
}

<ExampleWrapper>
  <TestCommonSwitch />
</ExampleWrapper>;
```
