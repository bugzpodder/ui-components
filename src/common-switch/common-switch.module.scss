@import "../styles/colors.scss";

.commonSwitch {
  display: inline-block;
  position: relative;

  .controlLabel {
    margin: 0;
  }

  .switchHelperText {
    margin-top: -7px;
    position: absolute;
  }

  .error {
    color: $grail-red-color;
  }
}
