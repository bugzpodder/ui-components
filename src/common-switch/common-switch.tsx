import FormControlLabel from "@material-ui/core/FormControlLabel";

import FormHelperText from "@material-ui/core/FormHelperText";
import React, { ChangeEvent, useState } from "react";
import Switch from "@material-ui/core/Switch";
import classNames from "classnames";
import styles from "./common-switch.module.scss";
import { CommonSwitchClasses } from "../types/switch";

type Props = {
  /** Returns the status of the switch, and the value if one exists */
  onChange: (x0: boolean, x1: any) => any;
  /** Determines the color of the switch. Options include "primary" and "secondary" */
  color?: "primary" | "secondary";
  /** The id given to the switch */
  id?: string;
  /** When `true`, displays the switch in an error state */
  showError?: boolean;
  /** Provides a helper text under the switch */
  helperText?: string;
  /** Provides a label to the right of the switch */
  label?: string;
  /** When `false`, disables the switch */
  isEnabled?: boolean;
  /** Allows the programmer to take control of the selected state of the switch */
  isSelected?: boolean;
  /** The value of the switch input if one is provided */
  value?: string;
  /**
   * Provides classNames to the switch's sub-components. Options include:
   *
   *  - `root`: switch's outermost div
   *
   *  - `label`: switch label
   */
  classes?: CommonSwitchClasses;
};

/** `CommonSwitch` renders a Material-UI switch component */
export const CommonSwitch: React.FC<Props> = props => {
  const [isChecked, setIsChecked] = useState(false);

  const {
    id = "",
    classes = {},
    color = "primary",
    showError = false,
    helperText = "",
    label,
    isEnabled = true,
    isSelected,
    value = "",
    onChange,
    ...controlProps
  } = props;

  const handleChange = (e: ChangeEvent<HTMLInputElement>, checked: boolean) => {
    onChange(checked, e.currentTarget.value);
    if (isSelected === undefined || isSelected === null) {
      setIsChecked(checked);
    }
  };

  const errorClass = { [styles.error]: showError };
  const switchSelection =
    isSelected === undefined || isSelected === null ? isChecked : isSelected;
  return (
    <div className={classNames(styles.commonSwitch, classes.root)}>
      <FormControlLabel
        classes={{
          root: styles.controlLabel,
          label: classNames(classes.label, errorClass),
        }}
        data-testid="common-switch"
        control={
          <Switch
            id={id}
            checked={switchSelection}
            classes={{
              checked: classNames(errorClass),
              colorPrimary: classNames(errorClass),
              colorSecondary: classNames(errorClass),
              track: classNames({ [styles.errorBar]: showError }),
            }}
            color={color}
            disabled={!isEnabled}
            disableRipple
            onChange={handleChange}
            value={value}
          />
        }
        label={label || null}
        {...controlProps}
      />
      {helperText && (
        <FormHelperText
          classes={{
            root: styles.switchHelperText,
          }}
          error={showError}
        >
          {helperText}
        </FormHelperText>
      )}
    </div>
  );
};
