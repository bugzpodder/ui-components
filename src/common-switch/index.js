// @flow
export { CommonSwitch } from "./common-switch";
