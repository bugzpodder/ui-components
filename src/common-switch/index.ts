export { CommonSwitch } from "./common-switch";
