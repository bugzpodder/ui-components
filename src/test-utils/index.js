// @flow
export {
  ALPHABET_ANIMALS_DATA,
  ADAMS_DATA,
  COUNTRIES,
  EMU_DATA,
  EXTRA_LARGE_EXAMPLE_TABLE_DATA,
  LARGE_EXAMPLE_TABLE_DATA,
  EXAMPLE_TABLE_DATA,
  GRAIL,
} from "./constants";
export { ExampleBlock } from "./example-block";
export { TestWrapper } from "./test-wrapper";
export { StyleWrapper as ExampleWrapper } from "../style-wrapper";
