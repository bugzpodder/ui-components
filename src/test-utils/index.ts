export {
  ALPHABET_ANIMALS_DATA,
  ADAMS_DATA,
  COUNTRIES,
  COUNTRY_NAMES,
  EMU_DATA,
  EXTRA_LARGE_EXAMPLE_TABLE_DATA,
  LARGE_EXAMPLE_TABLE_DATA,
  EXAMPLE_TABLE_DATA,
  GRAIL,
  TEST_EXTERNAL_DOMAINS,
} from "./constants";
export { ExampleBlock } from "./example-block";
export { TestWrapper } from "./test-wrapper";
export { StyleWrapper as ExampleWrapper } from "../style-wrapper";
