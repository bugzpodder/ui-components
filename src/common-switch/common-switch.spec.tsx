import React from "react";
import { CommonSwitch } from "./common-switch";
import { TestWrapper } from "../test-utils";
import { cleanup, fireEvent, render } from "@testing-library/react";

afterEach(cleanup);

test("render primary common switch", () => {
  const mockChange = jest.fn(result => result);
  const { container, getByTestId } = render(
    <TestWrapper>
      <div>
        <CommonSwitch
          label="Primary"
          helperText="test switch"
          value="test"
          onChange={mockChange}
        />
      </div>
    </TestWrapper>,
  );
  fireEvent.click(getByTestId("common-switch"));
  expect(mockChange).toHaveBeenCalledTimes(1);
  expect(mockChange.mock.results[0].value).toEqual(true);
  fireEvent.click(getByTestId("common-switch"));
  expect(mockChange).toHaveBeenCalledTimes(2);
  expect(mockChange.mock.results[1].value).toEqual(false);
  expect(container).toMatchSnapshot();
});

test("render secondary common switch", () => {
  const mockChange = jest.fn(result => result);
  const { getByTestId } = render(
    <TestWrapper>
      <div>
        <CommonSwitch
          isEnabled
          isSelected
          color="secondary"
          onChange={mockChange}
        />
      </div>
    </TestWrapper>,
  );
  fireEvent.click(getByTestId("common-switch"));
  expect(mockChange).toHaveBeenCalledTimes(1);
  expect(mockChange.mock.results[0].value).toEqual(false);
});

test("render error common switch", () => {
  const mockChange = jest.fn(result => result);
  const { container } = render(
    <TestWrapper>
      <div>
        <CommonSwitch
          showError
          color="secondary"
          isSelected
          onChange={mockChange}
        />
      </div>
    </TestWrapper>,
  );
  expect(container).toMatchSnapshot();
});
