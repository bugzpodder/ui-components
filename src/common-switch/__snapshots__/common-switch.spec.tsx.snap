// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render error common switch 1`] = `
<div>
  <div>
    <div
      class="commonSwitch"
    >
      <label
        class="MuiFormControlLabel-root controlLabel"
        data-testid="common-switch"
      >
        <span
          class="MuiSwitch-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-155 MuiSwitch-switchBase MuiSwitch-colorSecondary error PrivateSwitchBase-checked-156 Mui-checked error"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input-158 MuiSwitch-input"
                id=""
                type="checkbox"
                value=""
              />
              <span
                class="MuiSwitch-thumb"
              />
            </span>
          </span>
          <span
            class="MuiSwitch-track errorBar"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label error MuiTypography-body1"
        />
      </label>
      
    </div>
  </div>
</div>
`;

exports[`render primary common switch 1`] = `
<div>
  <div>
    <div
      class="commonSwitch"
    >
      <label
        class="MuiFormControlLabel-root controlLabel"
        data-testid="common-switch"
      >
        <span
          class="MuiSwitch-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-19 MuiSwitch-switchBase MuiSwitch-colorPrimary"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                class="PrivateSwitchBase-input-22 MuiSwitch-input"
                id=""
                type="checkbox"
                value="test"
              />
              <span
                class="MuiSwitch-thumb"
              />
            </span>
          </span>
          <span
            class="MuiSwitch-track"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Primary
        </span>
      </label>
      <p
        class="MuiFormHelperText-root switchHelperText"
      >
        test switch
      </p>
    </div>
  </div>
</div>
`;
